[WIP] Ajoute configuration de développement pour docker-compose

motet-a opened the pull request:
Les images se basent sur Debian et pas sur Alpine suite à quelques
problèmes de paquets binaires pas distribués pour musl.

Le conteneur du frontend a un accès en lecture seule aux sources, dans
le dossier `assets/`. Il fait tourner uniquement le processus de Gulp,
qui observe tous les fichiers et met à jour un volume monté sur
`dist/` dès le moindre changement.

Ce même volume `dist/` est monté en lecture seule dans le conteneur du
backend, qui fait tourner le processus de Django. Le dossier `zds/`
est également partagé en lecture seule.

La BDD, les fichiers des tutos et ceux des utilisateurs sont dans un
volume nommé, ils persistent.

Il n’y pas de LaTeX.

Il va falloir aussi rajouter un service pour le serveur de zmarkdown par la suite.

-------

Ajoute `clem`, script pour utiliser docker-compose facilement

Je pense que c’est l’idéal pour des gens qui veulent débugger
rapidement un bout de JS sans avoir à lire un tas de documentation
sur comment se servir de Docker.

Ajoute également une petite page de documentation là-dessus.

-----

Notez que j’utilise tout ceci depuis le début, je n’ai d’ailleurs jamais installé ZdS en “dur” sur mon PC. Cette PR a beaucoup évoluée depuis que je l’ai ouverte, il y a maintenant `clem`, de la documentation, des images plus légères et un support d’Elasticsearch.

----

- [ ] Il y a un petit problème quand on arrête `clem` avec un Ctrl+C. Ça s’arrête mais `docker-compose down` reste attaché bizarrement au terminal.

### Contrôle qualité

Vous devez [installer Docker et docker-compose](https://docs.docker.com/compose/install/).

1. Lancer `./clem up` et attendez (longtemps)
2. http://localhost:8000/
3. Il n’y a pas de troisème étape.

Consultez la documentation (ou `clem --help`) et testez le reste (Elasticsearch, fixtures, tests unitaires, …).

motet-a committed:
Ajoute configuation de développement pour docker-compose

Les images se basent sur Debian et pas sur Alpine suite à quelques
problèmes de paquets binaires pas distribués pour musl.

Le conteneur du frontend a un accès en lecture seule aux sources, dans
le dossier `assets/`. Il fait tourner uniquement le processus de Gulp,
qui observe tous les fichiers et met à jour un volume monté sur
`dist/` dès le moindre changement.

Ce même volume `dist/` est monté en lecture seule dans le conteneur du
backend, qui fait tourner le processus de Django. Le dossier `zds/`
est également partagé en lecture seule.

Pendant que ça tourne, les fixtures peuvent être appliquées comme ça :

    docker-compose exec back \
        python manage.py loaddata \
            --settings zds.settings_docker_dev \
            fixtures/users.yaml

Et lorsque tout est arrêté, les tests peuvent être lancés comme ça :

    docker-compose run back \
        python manage.py test \
            --settings zds.settings_docker_test

Et pour lint le backend :

    docker-compose exec back \
        make lint-back

La BDD, les fichiers des tutos et ceux des utilisateurs sont dans un
volume nommé, ils persistent.

pierre-24 commented:
Ping @sandhose qui avait déjà un peu joué avec ça :)

DevHugo commented on docker-compose.yml:
```diff
@@ -0,0 +1,34 @@
+      - ./quotes.txt:/zds/quotes.txt:ro
+      - ./scripts:/zds/scripts:ro
+      - ./templates:/zds/templates:ro
+      - ./zds:/zds/zds:ro
```
quand on créé des contenu ou qu'on upload des images ou archives, on va avoir besoin des droits sur le dossier en écriture, non ?

motet-a replied:
Si j’ai bien compris, les tutoriels, les images et tout sont dans des dossiers à la racine du dépot. Du coup, comme ce n’est dans aucun volume, il y a un accès en lecture-écriture depuis le conteneur.

https://github.com/zestedesavoir/zds-site/blob/dev/zds/settings.py#L482

Par contre ils ne persisteront pas. Je pense qu’il faudrait les sauvegarder dans un volume comme la BDD dans l’idéal.

pierre-24 replied:
Si ils ne persistent pas alors que la BDD elle persiste, effectivement, il va y avoir des suprises (en particulier pour les contenus). Donc ce serait intéréssant, oui.

DevHugo replied:
Pierre, a raison, c'est le point que je voulais soulever mais les armées de manchots ont été plus rapide.

DevHugo reviewed (approved):
Manque quand même un minimum de documentation, au moins une procédure d'installation comme on a fait pour le back ou le front.

motet-a commented:
@DevHugo Je suis tout à fait d’accord, il manque de la documentation. Mais je pense que c’est encore un peu trop top pour en écrire vu que pour l’instant, c’est surtout un bidouillage qui ne sert qu’à moi et que l’arrivée de zmarkdown va tout chambouler.

pierre-24 commented:
Pas d'accord. On nous a repproché par le passé la lourdeur de notre procédure d'installation, et c'est une solution possible. Même si ton alternative est "expérimentale", je pense que ça vaudrait la peine de la documenter, quitte à noter "expérimental" :p 

(puis au pire, tu fait une PR sur la branche du nouveau zmd)

DevHugo commented on Dockerfile_front:
```diff
@@ -0,0 +1,12 @@
+
+RUN npm install
+
+CMD ["npm", "run", "gulp"]
```
manque pas un mot clé comme watch ?

motet-a replied:
Non, c’est bien ça : https://github.com/zestedesavoir/zds-site/blob/dev/package.json#L9

Ça observe bien les changements.

(Je reconnais que le nom est un peu trompeur)

DevHugo replied:
Si tu es sûr de toi, pas de souci mais j'avais l'impression que ça fera qu'un build au début et que c'est différent de watch.

motet-a replied:
C’est bien trompeur mais c’est pas le cas :-D

DevHugo commented on Dockerfile_back:
```diff
@@ -0,0 +1,19 @@
+FROM buildpack-deps:stretch
+
+RUN apt-get update \
```
Es-ce pertinent de le faire au début ? tu va empêcher docker de mettre en cache ton image.

motet-a replied:
Si c’est à propos du `apt-get update`, j’ai toujours vu ça [fait comme ça](https://github.com/docker-library/php/blob/ddc7084c8a78ea12f0cfdceff7d03c5a530b787e/7.1/Dockerfile#L21).

DevHugo replied:
L'instruction `apt-get update` est correcte et relativement indispensable. Elle permet de mettre à jours les paquets de ta distribution.

Je parlais de la suite d'instruction de 3 à 10. La question est en gros quelles conséquences cela a sur ton cache, si tu change ne serait qu'un caractère sur les lignes 4 à 8. N'as tu pas intérêt à organiser tes paquets sur une ligne et de créer des layers en fonctions des paquets qui peuvent ne plus être utilisé et ceux qui ne seront jamais modifié ?

motet-a replied:
Ça peut en effet être très intéressant, mais il faudrait demander à un connaisseur du backend pour ça.

pierre-24 replied:
Ce genre de paquet n'est pas amené à changer, sauf installation d'un nouvel outil (et encore). Les changements se font en général dans les `requirements*.txt`.

DevHugo commented on scripts/docker_dev_entrypoint.sh:
```diff
@@ -0,0 +1,8 @@
+cd ..
+
+python manage.py migrate --settings zds.settings_docker_dev
+python manage.py runserver 0.0.0.0:8000 --settings zds.settings_docker_dev
```
Y'a surement un moyen de mettre le port dans le docker-compose, c'est important car le port 8000 peut-être utilisé sur la machine du développeur. 

Pouvoir le changer simplement est utile et intéréssant.

motet-a replied:
Ce point d’entrée est toujours exécuté à l’intérieur du conteneur par docker-compose. Et à l’intérieur du conteneur, le port 8000 sera forcément libre.

Le changement de port doit se faire [ici](https://github.com/zestedesavoir/zds-site/pull/4430/files/ec208d1563cf3107dbfa21e57911f235ee5e2379#diff-4e5e90c6228fd48698d074241c2ba760R31), dans le `docker-compose.yml`. Pour l’instant le port est en dur mais je crois qu’il est possible d’utiliser une variable d’environement avec une valeur par défaut. Je changerai ça.

DevHugo replied:
ok. variable d’environement avec une valeur par défaut me convient.

motet-a replied:
C’est fait, maintenant on peut même faire `./clem up --address 127.0.0.1 --port 8000`.

DevHugo commented:
> (puis au pire, tu fait une PR sur la branche du nouveau zmd)

ou un sujet sur le forum. Je suis entièrement d'accord avec pierre-24, on ne pourra pas merger sans documentation.

Question: du coup, qui va builder les images, le développer, dans ton cas. As-tu mesurer les avantages et inconvénients de ta solution par-rapport un build côté travis et téléchargement ?

pierre-24 commented:
> Question: du coup, qui va builder les images, le développer, dans ton cas. As-tu mesurer les avantages et inconvénients de ta solution par-rapport un build côté travis et téléchargement ?

Même si ça serait assez intéréssant d'avoir par exemple une image pour chaque PR, j'ai peur que comme on part sur du "debian bien lourd", l'image sois pas légère. Mais je parle sans savoir ^^

DevHugo commented:
> Même si ça serait assez intéréssant d'avoir par exemple une image pour chaque PR, j'ai peur que comme on part sur du "debian bien lourd", l'image sois pas légère. Mais je parle sans savoir ^^

Je pose la question, c'est tout ^^, ma question n'est pas orienté.

motet-a commented:
Merci pour vos reviews. J’ai rajouté [WIP] au titre de la PR, je suis en train de m’occupper du problème de documentation et surtout de simplicité d’utilisation du bazar.

Concernant le poids des images de Debian, c’est tout relatif. Compressée, [l’image de Debian 8](https://hub.docker.com/r/library/debian/tags/8/) ne fait que 50 Mio, et ça monte vers 100 Mio décompressé. C’est juste que [les images d’Alpine](https://hub.docker.com/r/library/alpine/tags/) sont minuscules.

DevHugo commented:
> Concernant le poids des images de Debian, c’est tout relatif. Compressée, l’image de Debian 8 ne fait que 50 Mio, et ça monte vers 100 Mio décompressé. C’est juste que les images d’Alpine sont minuscules.

On vera sur combien ça joue sur le poids final de ton image, si c'est négligeable ou pas. Combien de temps prends ton image à build ? as tu vérifié la sécurité de ton image avec les outils associés sur docker hub ?

motet-a commented:
> On vera sur combien ça joue sur le poids final de ton image, si
  c'est négligeable ou pas.

Ça joue assez peu sur la taille des images finales. Actuellement,
celle du backend fait 1 Gio et celle du frontend 755 Mio. C’est
pas tout petit mais je ne sais pas si ça vaut la peine de faire
plus d’effort pour réduire ça.

> Combien de temps prends ton image à build ?

Chez moi, ça prends un petit bout de temps parce que je suis en
ADSL. Mais personellement, je préfère patienter 15 minutes qu’une
image build en faisant autre chose que de passer 30 minutes à
lire de la doc pour faire tourner ce site chez moi :-)

> As tu vérifié la sécurité de ton image avec les outils associés
  sur docker hub ?

Non. Mais étant donné que ce n’est pas pour la production, je
doute que ce soit important.

DevHugo commented:
> Ça joue assez peu sur la taille des images finales. Actuellement,
celle du backend fait 1 Gio et celle du frontend 755 Mio. C’est
pas tout petit mais je ne sais pas si ça vaut la peine de faire
plus d’effort pour réduire ça.

ça me parait beaucoup, les images que j'avais build à l'époque (y'a un an) : 
 - [front-latest 146 MB](https://hub.docker.com/r/arhenion/zds-docker/)
 - [back-latest 154 MB](https://hub.docker.com/r/arhenion/zds-docker/)

Peut-être qu'on parle pas de la même chose.

> Chez moi, ça prends un petit bout de temps parce que je suis en
ADSL. Mais personellement, je préfère patienter 15 minutes qu’une
image build en faisant autre chose que de passer 30 minutes à
lire de la doc pour faire tourner ce site chez moi :-)

Le tout est de savoir étape par étape le temps de build qu'on puisse optimiser tout cela, par exemple, comment s'organise t-on pour faire le npm install, qui le fait et pourquoi.
Je ne conteste pas le choix de docker mais je pose des questions pour qu'on puisse au mieux cet outil. 

> Non. Mais étant donné que ce n’est pas pour la production, je
doute que ce soit important.

C'est vrai, n'empêche qu'on peut appliquer des principes de sécurités simple et peut couteux en terme de développement, par-exemple, as tu besoin d'être root sur ton container ?
C'est pas indispensable, c'est juste pour soulever la question.

motet-a commented:
> ça me parait beaucoup, les images que j'avais build à l'époque (y'a un an)

Les tailles des images indiquées sur le Docker Hub sont compressées. Décompressée, `arhenion/zds-docker:front-latest` fait 441 Mio, contrairement à la mienne qui fait 755 Mio. La mienne se base sur `node:8` qui se base sur du `buildpack-deps`, elle inclus donc GCC et compagnie. Je crois que j’avais des problèmes avec `node:8-slim` parce que certains packets de Node.js étaient dépassés et ne fournissaient pas de binaires déjà compilés à télécharger. Mais maintenant que les deps sont à jour, il faut que j’essaye de changer ça. Si on se base sur du `node:slim`, le temps de construction devrait être vraiment réduit vu qu’il n’y aura aucune compilation lors du `npm install`. 

J’ai pas regardé le backend en détails. Ça serait bien d’avoir un retour de quelqu’un qui connait bien Django. Je connais mieux Node.js que Django :upside_down_face: 

> Le tout est de savoir étape par étape le temps de build qu'on puisse optimiser tout cela

J’ai pas une connection à internet très représentative actuellement et je pense qu’il faudrait déjà bien s’assurer que ça marche correctement, de corriger les problèmes évidents et d’écrire la doc avant d’optimiser.

> C'est vrai, n'empêche qu'on peut appliquer des principes de sécurités simple et peut couteux en terme de développement, par-exemple, as tu besoin d'être root sur ton container ?

Je ne suis pas spécialiste mais je ne vois pas l’intérêt de ne pas être root dans ce cas. Les conteneurs Docker sont complètement isolés entre eux et de l’hôte. Si je suis root dans un conteneur, j’ai un uid invalide par rapport à l’hôte.

motet-a commented:
Btw, `arhenion/zds-docker:back-latest` fait curieusement que 379 Mio décompressée. Il faut que j’essaye cette image.

motet-a committed:
Ajoute `clem`, script pour utiliser docker-compose facilement

Bon, j’étais un peu en panne d’inspiration pour le nom, du coup il
s’appelle `clem` pour l’instant. C’est un petit script compatible avec
Python 2 et Python 3 qui lance docker-compose.

Évidemment, c’est très limité. Mais je pense que c’est l’idéal pour
des gens qui veulent débugger rapidement un bout de JS sans avoir à
lire un tas de documentation sur comment se servir de Docker.

Ajoute également une petite page de documentation là-dessus.

pierre-24 commented on clem:
```diff
@@ -0,0 +1,185 @@
+    test_front_parser.set_defaults(func=test_front)
+
+
+create_test_parser()
```
J'avoue que j'ai du mal avec la logique de la fonction qui manipule des variables globale et n'est appellée qu'une fois dans le script. J'ai raté un truc, ou bien ? ^^

motet-a replied:
Je vais remplacer ça par un `if True:` comme me l’a dit vhf.

motet-a replied:
D’ailleurs, les variables globales ne sont pour la plupart assignées qu’une seule fois, comme avec `const` en JS ou `final` en Java. C’est pas trop sale comme ça.

pierre-24 commented on doc/source/index.rst:
```diff
@@ -14,6 +14,7 @@ Sommaire
 .. toctree::
    :maxdepth: 2
 
+   docker
```
Je l'aurais mis dans `install` : c'est une des manière de faire tourner zds.

pierre-24 replied:
(ou dans `utils`)

motet-a replied:
Ouais j’hésitais. Comme ça parle aussi de comment faire tourner ZdS, des migrations, j’ai pensé que ça couvrait d’autres choses que l’installation. Mais c’est peut-être mal.

pierre-24 replied:
Ben là, à la racine, ça lui donne une importance démesurée par rapport au reste, en fait ^^

pierre-24 commented on doc/source/docker.rst:
```diff
@@ -0,0 +1,100 @@
+
+.. __: https://docs.docker.com/compose/install/
+
+Ensuite, `installez Git`__. Si vous êtes sous Linux ou Mac OS X, il
```
Cette partie (et la partie clonage) n'ont pas besoin d'être explicités plus que ça, elle le sont déjà dans la partie "installer zds"). Ne te casse pas la tête :-)

motet-a replied:
C’est pas faux.

pierre-24 commented on doc/source/docker.rst:
```diff
@@ -0,0 +1,100 @@
+Appuyez sur Ctrl+C pour tout arrêter.
+
+
+Liste des commandes principales de ``clem``
```
Je crois me souvenir que le nombre de `-` doit être égal au nombre de letrres dans le titre.

motet-a replied:
Oups.

motet-a commented:
Bon, il y a un gros problème avec les settings en fait, je suis en train de régler ça sur la #4438.

pierre-24 commented on doc/source/docker.rst:
```diff
@@ -0,0 +1,100 @@
+----------------
+
+Premièrement, `installez Docker et docker-compose`__ (vous aurez
+besoin des deux).
```
... Par contre, au vu de ce qui se fait ailleurs dans la doc, tu peux être plus explicite ici. Même s'il suffit probablement de metrre qu'un `apt-get` suffit. Voir [documentation existante](http://zds-site.readthedocs.io/fr/latest/install.html) (la philosophie, c'est de mettre le strict nécéssaire, mais de le mettre :p )

motet-a replied:
C’est vrai.

Surtout qu’il ne suffit pas de faire un simple `apt-get` malheureusement : Sur Debian, Fedora et autres, le paquet `docker` du dépot officiel de la distro est souvent assez vieux et la doc de Docker dit d’ajouter leurs dépot d’APT histoire d’avoir une version bien à jour.

motet-a replied:
Mais je me baserai bien sur de la doc déjà existante pour ça. Ça va être relou à maintenir par nous et les procédures d’installation de Docker risquent de changer.

pierre-24 replied:
Je t'en prie :) (j'ai une VM sous windows pour tester, par contre le soucis, c'est toujours de trouver un maceux ... @DevHugo ? :-° )

motet-a replied:
@pierre-24 Oh, normalement il ne devrait pas y avoir de problèmes sous macOS, c’est surtout Windows qui me fait peur avec Powershell, WSL et tout le bazar.

DevHugo replied:
Ouai, je peux tester, pas le temps de lire tous l'échange.

motet-a commented:
Bon, ne prenez pas peur, il y a dans les commits des choses qui n’ont rien à voir avec cette PR mais ils seront enlevés. Il faut juste ne pas merge cette PR pour l’instant.

motet-a commented:
Tous les commits sans rapport sur lesquels était basée cette PR ont étés merge dans `dev` !

motet-a commented:
Je viens de mettre à jour la documentation [que vous pouvez voir en ligne ici](http://doc.zds.3.141.ovh/install/docker.html).

motet-a committed:
On passe à yarn

motet-a committed:
Permet de lancer les tests avec selenium et Firefox

Utilise Xvfb un peu comme Travis.

motet-a committed:
Python 3

motet-a committed:
Préférrer DJANGO_SETTINGS_MODULE à --settings

Ça fait moins de répétitions.

motet-a committed:
Ne pas demander la confirmation de suppression des conteneurs

AmarOk1412 commented:
Pour le moment j'obtiens :

[temp.txt](https://github.com/zestedesavoir/zds-site/files/1600134/temp.txt)


je regarderais plus tard

AmarOk1412 committed:
docker-compose: fix les options pour monter les volumes

L'utilisation de SELinux pour Docker (active sur les distributions type RHEL)
empechera l'utilisation correcte des volumes. Le flag :z doit-etre mis sur les
fichiers utilises par de multiples containers (ou :Z si un seul container
utilise le fichier).

Eskimon commented on Dockerfile_back:
```diff
@@ -0,0 +1,27 @@
+FROM python:3-stretch
```
Faudrait sûrement préciser une version similaire à celle de la prod', pour se rapprocher des conditions de prod'.

Eskimon commented:
Je trouve cette PR intéressante, ca serait cool qu'elle voit le jour. Il y aurait moyen d'ajouter un service mysql pour avoir une bdd proche de celle de la prod' (et tester ainsi plus facilement les cas limites liés aux différences mysql/sqlite) ?

firm1 commented:
Bonjour,

Vu le nombre de PR en cours, j'aimerai faire un peu de QA pour vider un peu la stack. C'est la raison qui m'amène sur cette PR. Je ne sais pas si @motet-a est encore actif, car j'aimerai bien reprendre sa PR pour l'amener au bout. J'ai noté les éléments suivants qui sont a faire : 

- passer sur le nouveau zmd (ce n'est pas encore le cas dans cette PR)
- passer sur mysql au lieu de sqlite comme le suggère @Eskimon ci-dessus
- mettre à jour la documentation

Voilà un peu le plan que j'ai. ça vous convient ?

firm1 commented:
Je ferme ici, on reprend les choses [là bas](https://github.com/zestedesavoir/zds-site/pull/5012).
